### PR TITLE
[FLINK-19775][test] SystemProcessingTimeServiceTest.testImmediateShutdown is instable

### PR DESCRIPTION
**Pull request by xiaoHoly**
## What is the purpose of the change

I will repair the test function (SystemProcessingTimeServiceTest.testImmediateShutdown)  which  is instable .
Express in the FLINK-19775

## Brief change log

-change flink/flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/OneShotLatch.java

## Verifying this change

This change is already covered by existing tests, such as SystemProcessingTimeServiceTest.testImmediateShutdown.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (no)

### CODE REVIEW

**Comment by xiaoHoly** [comment]
@dianfu ,hi,Teacher Fu,I have finished my work for this problem . Can you help me review my code?

**Comment by xiaoHoly** [comment]
@flinkbot run travis re-run the last Travis build

**Comment by xiaoHoly** [comment]
@flinkbot run azure re-run the last Azure build

**Comment by xiaoHoly** [comment]
@flinkbot run travis re-run the last Travis build

**Comment by xiaoHoly** [comment]
@flinkbot run azure re-run the last Azure build

**Comment by dianfu** [comment]
@xiaoHoly It will trigger the test atomically when you pushing a new commit. Usually there is no need to trigger the test manually. Besides, you can use command `@flinkbot run azure` instead of `run azure re-run the last Azure build` to trigger the tests.

**Comment by dianfu** [comment]
I'm not quite familiar with this part.
cc @tillrohrmann

**Comment by xiaoHoly** [comment]
> @xiaoHoly It will trigger the test atomically when you pushing a new commit. Usually there is no need to trigger the test manually. Besides, you can use command `@flinkbot run azure` instead of `run azure re-run the last Azure build` to trigger the tests.

Thanks for your advice

**Comment by xiaoHoly** [comment]
HI,@tillrohrmann .PTAL

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/OneShotLatch.java [hunk 1]
```diff
@@ -61,7 +61,11 @@ public void await() throws InterruptedException {
 				try {
 					waitersSet.add(thread);
 					lock.wait();
-				} finally {
+				}
+				catch (InterruptedException e) {
+					Thread.currentThread().interrupt();
+				}
+				finally {
```
Comment (tillrohrmann) [inline flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/OneShotLatch.java:68]: I don't understand what this change will do for fixing the `SystemProcessingTimeServiceTest` instability.

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks for creating this PR @xiaoHoly. I don't understand how the change you introduced will fix the test instability. Moreover, I believe the previous implementation of `OneShotLatch.await` was correct.

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/OneShotLatch.java (hunk 1 above)
Comment (xiaoHoly) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/OneShotLatch.java:68]: hi ,@tillrohrmann.I'm not sure if I can solve the instability with the change,but it does exist.

Comment (tillrohrmann) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/OneShotLatch.java:68]: Without an explanation why this solves the problem, I cannot merge this PR.

Comment (xiaoHoly) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/OneShotLatch.java:68]: ok,thanks.I will spend more time researching this issue

Comment (tillrohrmann) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/OneShotLatch.java:68]: Ok, then let's close this PR for the time being.

**Comment by tillrohrmann** [comment]
Closing this PR until the root cause of the problem has been found.

**Comment by xiaoHoly** [comment]
@tillrohrmann ,Maybe I found the reason for the instability. It depends on when the thread A calling the await() method acquires the lock of the lock object in the jvm lock pool. According to the lock.wait() code in the await() method, we will know that it will cause the current thread A to enter the blocking state. At this point, trigger() will perform its work in the thread B to which it belongs. Suppose that when thread B executes to lock.notifyAll(), thread A will be awakened, but the lock of the lock object is not allocated in time and is still blocked. As long as the interrupt flag  of thread A is true, InterruptedException will be thrown. So I thought of two solutions. The first point: Thread B changes the interrupt flag  of thread A to false before executing lock.notifyAll(). This method can ensure that the A thread will not throw InterruptedException after being awakened, but how to restore the thread interruption flag for different threads still needs to be considered. The second point: Use catch code to catch the exception in the A thread, it will automatically restore the thread interruption flag, and make the compilation pass. But the second method cannot solve the problem of exception throwing.

**Comment by xiaoHoly** [comment]
To facilitate the discussion, I will post the code below：

method of  Thead A!
public void await() throws InterruptedException {
		synchronized (lock) {
			while (!triggered) {
				Thread thread = Thread.currentThread();
				try {
					waitersSet.add(thread);
					lock.wait();
				} finally {
					waitersSet.remove(thread);
				}
			}
		}
	}

method of  Thead B!
	public void trigger() {
		synchronized (lock) {
			triggered = true;
			lock.notifyAll();
		}
	}

**Pull request closed without merging**